Deprecate disabling basic-license features

williamrandolph opened the pull request:
We believe there's no longer a need to be able to disable basic-license features completely using the `xpack.*.enabled` settings. If users don't want to use those features, they simply don't need to use them. Having such features always available lets us build more complex features that assume basic-license features are present.

This commit deprecates settings of the form "xpack.*.enabled" for basic-license features, excluding "security", which is a special case.

Meta issue: #54745

williamrandolph committed:
Deprecate disabling basic-license features

We believe there's no longer a need to be able to disable basic-license
features completely using the "xpack.*.enabled" settings. If users don't
want to use those features, they simply don't need to use them. Having
such features always available lets us build more complex features that
assume basic-license features are present.

This commit deprecates settings of the form "xpack.*.enabled" for
basic-license features, excluding "security", which is a special case.

williamrandolph committed:
Account for deprecation warnings in tests

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Plugins)

williamrandolph committed:
Remove deprecated settings from testclusters

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph committed:
Catch deprecation warning in enrich test

williamrandolph committed:
Apply spotless formatting

williamrandolph committed:
Move deprecation handling to, er, correct test

mark-vieira commented:
> However, we are planning to make the breaking change of removing those settings altogether on the master branch and I will remove one of those test tasks when I make the breaking changes.

Yes, we need to keep the coverage if it's just deprecated but if this is slated for _removal_ in `8.0` then I _think_ we can ditch this project in `master`. @rjernst does this seem reasonable? I'm not _completely_ sure what we are testing in terms of licensing here that needs to stick around or if it's even related to this PR.

[Here's the tests](x-pack/qa/rolling-upgrade-basic/src/test/java/org/elasticsearch/upgrades/BasicLicenseUpgradeIT.java) specific to this project.

williamrandolph committed:
Revert "Remove deprecated settings from testclusters"

This reverts commit 92e909ac88b3c6a6b917e7944bbd89eacf1a3d82.

williamrandolph committed:
Handle setting deprecation warnings in tests

williamrandolph committed:
Handle deprecation warnings in Watcher tests

williamrandolph committed:
Apply spotless formatting

williamrandolph committed:
Remove unused imports

williamrandolph committed:
Handle deprecations in CCR test

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

rjernst commented on x-pack/plugin/watcher/qa/rest/src/test/resources/rest-api-spec/test/mustache/30_search_input.yml:
```diff
@@ -13,6 +16,8 @@ setup:
             "value" : "val_1"
           }
+      warnings:
+        - "[xpack.ilm.enabled] setting was deprecated in Elasticsearch and will be removed in a future release! See the breaking changes documentation for the next major version."
```
why are we getting these warnings? Since this is testing ilm enabled, shoudln't the test setup be using the default and thus we get no warning? I think the problem is many of these qa tests are trying to disable other features, but we should simply remove those settings from build.gradle in the cluster setup?

williamrandolph replied:
I agree. I've been trying to identify the specific places where deprecation warnings are thrown in order to make sure that there's nothing important to the tests about having particular settings in `build.gradle`, but in the case of the watcher tests, the deprecation warnings seem to be happening on index creations and deletions, which doesn't seem relevant to watcher.

rjernst reviewed (commented):
I think the question of `x-pack:qa:rolling-upgrade-basic` vs `x-pack:qa:rolling-upgrade` is orthogonal to this issue. We should look at that separately. From what I remember the upgrade-basic test was added when we opened x-pack, but it really has nothing to do with whether the basic licensed features have the ability to be disabled.

williamrandolph committed:
Enable ILM in watcher tests

The Watcher REST integration tests had been running with ILM disabled.
With that setting's deprecation, it seems that some index creation and
deletion operations will throw deprecation warnings. These deprecation
warnings don't seem to have to do with the purposes of the tests, so we
will just change the settings in build.gradle in order to avoid having
to catch them all.

williamrandolph committed:
Restore default warning handling in watcher test suite

williamrandolph committed:
Remove redundant deprecated settings from ilm tests

williamrandolph committed:
Remove disable monitoring for ccr tests

williamrandolph committed:
Remove redundant ilm setting from tests

williamrandolph committed:
Handle more deprecations

williamrandolph committed:
Back out change to ESRestTestCase

williamrandolph committed:
Remove more qa disablings of ILM

williamrandolph committed:
Handle more deprecation warnings

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph committed:
Remove unused imports

williamrandolph committed:
More deprecations

williamrandolph committed:
Remove unused imports

williamrandolph committed:
Enable ILM for security tests

williamrandolph committed:
Remove disabling sql in testclusters

williamrandolph committed:
Enable ILM for qa tests

williamrandolph committed:
Enable ILM for qa tests part 2

williamrandolph committed:
Enable ILM for qa tests part 3

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph committed:
Add ILM named-writeables for reading cluster state

williamrandolph committed:
Enable Monitoring in qa test

williamrandolph committed:
Remove unused imports

williamrandolph commented:
I'm going to do the docs changes in this PR because there aren't very many of them. `xpack.enrich.enabled`, `xpack.rollup.enabled`, `xpack.flattened.enabled`, and `xpack.vectors.enabled` don't seem to have ever been documented in the first place.

williamrandolph committed:
Update docs for deprecated settings

williamrandolph commented:
@debadair I'm deprecating some settings and updating the docs to match. I've tried to follow the examples in the docs guide but I would welcome advice on how to improve things here.

williamrandolph commented:
@elasticmachine run elasticsearch-ci/2

debadair commented on docs/reference/migration/migrate_8_0/settings.asciidoc:
```diff
@@ -51,3 +51,17 @@ setting `cluster.remote.connect` is removed.
 the `node.local_storage` setting has been removed.
+
+[float]
+==== Option to disable basic license features is removed
```
```suggestion
==== Option to disable basic license features is deprecated
```

debadair commented on docs/reference/modules/node.asciidoc:
```diff
@@ -52,10 +52,9 @@ fails to start.
-you want to use {transforms}, there must be at least one {transform} node in
-your cluster. For more information, see <<transform-settings>> and
-<<transforms>>.
+A node that `node.transform` set to `true`. If you want to use {transforms},
```
```suggestion
A node that has `node.transform` set to `true`. If you want to use {transforms},
```

debadair commented on docs/reference/modules/node.asciidoc:
```diff
@@ -406,19 +396,18 @@ node.remote_cluster_client: false <9>
-If you want to use {transforms} in your cluster, you must have
-`xpack.transform.enabled` set to `true` on all master-eligible nodes and all
+If you want to use {transforms} in your cluster, you must not have
+`xpack.transform.enabled` set to `false` on any master-eligible nodes or any
```
```suggestion
`xpack.transform.enabled` set to `false` on any master-eligible or
```

debadair commented on docs/reference/modules/node.asciidoc:
```diff
@@ -406,19 +396,18 @@ node.remote_cluster_client: false <9>
 data nodes. If you want to use {transforms} in clients (including {kib}), it
-must also be enabled on all coordinating nodes. You must also have
+also must not be disabled on any coordinating nodes. You must also have
 `node.transform` set to `true` on at least one node. This is the default
```
The "it" in this sentence is a bit ambiguous. I'd say:

To use {transforms} in clients {including {kib}), do not disable this setting on any coordinating nodes.

debadair commented on docs/reference/settings/monitoring-settings.asciidoc:
```diff
@@ -28,6 +28,7 @@ For more information, see <<monitor-elasticsearch-cluster>>.
 
 `xpack.monitoring.enabled`::
 Set to `true` (default) to enable {es} {monitoring} for {es} on the node.
+deprecated:[7.8.0,Basic License features are always enabled]
```
The deprecation warnings should come before the descriptions.

debadair reviewed (changes requested):
Generally looks good, just a few things that should be tweaked.

williamrandolph committed:
Adjust docs in line with PR feedback

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph commented:
@debadair Thank you for the feedback and fixes!

debadair reviewed (approved):
LGTM

williamrandolph commented:
@elasticmachine please run elasticsearch-ci/2

williamrandolph commented:
I'm running `checkPart2` a few times on a cloud instance so I can be sure that the changes here don't slow down the test suite.

williamrandolph commented:
After letting `checkPart2` run on a debian-9 cloud instance, I've got timings of 50:20, 51:30, and 51:34 for this branch. Meanwhile, the same instance ran `checkPart2` on `master` in 52:44. While this is a pretty small sample size, it leads me to believe that these changes aren't going to slow down our test suite too much.

williamrandolph committed:
Add undocument settings to migration list

rjernst commented on x-pack/plugin/watcher/qa/with-security/build.gradle:
```diff
@@ -18,7 +18,6 @@ restResources {
 testClusters.integTest {
   testDistribution = 'DEFAULT'
-  setting 'xpack.ilm.enabled', 'false'
   setting 'xpack.monitoring.enabled', 'false'
```
Shouldn't this also be removed?

rjernst commented on x-pack/qa/multi-cluster-search-security/build.gradle:
```diff
@@ -24,7 +24,6 @@ testClusters.'remote-cluster' {
-  setting 'xpack.ilm.enabled', 'false'
   setting 'xpack.security.enabled', 'true'
   setting 'xpack.watcher.enabled', 'false'
   setting 'xpack.monitoring.enabled', 'false'
```
This as well, isn't monitoring.enabled deprecated here?

rjernst reviewed (approved):
Looks good. The main nit is there seems to be several qa tests that still have monitoring disabled.

williamrandolph committed:
Merge branch 'master' into deprecate-disabling-basic-features

williamrandolph committed:
Remove monitoring disablement from plugin tests

williamrandolph committed:
Remove ilm disablement from plugin tests

williamrandolph commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/monitoring/collector/enrich/EnrichStatsCollectorTests.java:
```diff
@@ -87,12 +89,16 @@ public void testShouldCollectReturnsFalseIfEnrichIsDisabled() {
             verify(licenseState).isMonitoringAllowed();
         }
+        if (isElectedMaster && isMonitoringAllowed) {
+            assertSettingDeprecationsAndWarnings(new Setting<?>[] { XPackSettings.ENRICH_ENABLED_SETTING });
```
I will add a comment for clarity.

williamrandolph committed:
Address PR nits
